### PR TITLE
Decouple Difficulty and Version Fields

### LINKED ISSUE #317: Split Unit code into properties
Currently the `Unit` model uses
```python
    code = models.CharField(
        max_length=255,
        help_text="The version code for the handout, like 'ZGX'",
    )
```
which is a bit of a hardcode, especially as subject is already specified by its unit group.
Splitting this field into a version choice field and a difficulty choice field might be better, and constructing the code can be done as a property.

### PR DESCRIPTION
**Pull request by Incompleteusern**
Fixes #317.

Implements version codes as
```python
DIFFICULTY_CHOICES = (
    ("B", "Easy/Bet"),
    ("D", "Medium/Dalet"),
    ("Z", "Hard/Zayin"),
)

VERSION_CHOICES = (
    ("V", "V (Secret)"),
    ("W", "W"),
    ("X", "X"),
    ("Y", "Y"),
)

difficulty = models.CharField(
    max_length=2,
    choices=DIFFICULTY_CHOICES,
    help_text="The difficulty code for the handout, like B",
)
version = models.CharField(
    max_length=2,
    choices=VERSION_CHOICES,
    help_text="The version code for the handout, like W",
)
```

Benefits of this include being less prone to typos and less redundant information stored as the group subject.

Other things this implies
1. Migrate the database. Probably make a backup before that? See the custom code in the migration.
2. There's likely some latent bugs here. Will fix as notified. Free 500 diamond I suppose.
3. venueQ needs to be updated for this
4. fixtures should be re-dumped to conform to this format. Maybe deprecate all.json, though not sure anymore.

### CODE REVIEW

File: core/templates/core/unit_list.html [hunk 1]
```diff
@@ -19,7 +19,8 @@
     <li>The second letter repeats the olympiad category.</li>
     <li>
       The third letter is a version identifier (either W, X, or Y).
-      Many topics have multiple versions so they can be repeated in different years.
+      Many topics have multiple versions so they can be repeated in different years,
+      but you can do the same topic multiple times in one year as well.
```
Comment (Incompleteusern) [inline core/templates/core/unit_list.html:23]: Added this comment since I misinterpreted it the first time.

File: rpg/tests.py [hunk 2]
```diff
@@ -115,7 +115,7 @@ def test_level_up(self):
         alice = self.get_alice()
         self.login(alice)
         bonus = BonusLevelFactory.create(group__name="Level 40 Quest", level=40)
-        bonus_unit = UnitFactory.create(group=bonus.group, code="DKU")
```
Comment (Incompleteusern) [inline rpg/tests.py:118]: Cursed U version code?

Comment (jatloe) [reply rpg/tests.py:118]: WHAT IS THIS

Comment (jatloe) [reply rpg/tests.py:118]: what's "Level 40 Quest"

Comment (Incompleteusern) [reply rpg/tests.py:118]: It refers to a secret unit level.

Comment (jatloe) [reply rpg/tests.py:118]: i prefer meow :3

File: core/migrations/0058_alter_unit_unique_together_unit_difficulty_and_more.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+# Generated by Django 4.2.5 on 2023-10-26 01:26
+
+from django.db import migrations, models
+
+
+def set_difficulty_and_version(apps, schema_editor):
+    # We can't import the Person model directly as it may be a newer
```
Comment (vEnhance) [inline core/migrations/0058_alter_unit_unique_together_unit_difficulty_and_more.py:7]: seems to be a copy paste error?

File: core/migrations/0058_alter_unit_unique_together_unit_difficulty_and_more.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            name="version",
+            field=models.CharField(
+                choices=[("W", "W"), ("X", "X"), ("Y", "Y")],
+                default="W",
+                help_text="The version code for the handout, like W",
+                max_length=2,
+            ),
+            preserve_default=False,
+        ),
+        migrations.RunPython(set_difficulty_and_version),
```
Comment (vEnhance) [inline core/migrations/0058_alter_unit_unique_together_unit_difficulty_and_more.py:48]: doesn't RunPython need a reverse operation argument or something like that? at least i feel like i've had to specified noop a bunch before in the past

Comment (Incompleteusern) [reply core/migrations/0058_alter_unit_unique_together_unit_difficulty_and_more.py:48]: I think that its optional? Stackoverflow had that at some point and it seems to migrate properly.

**Comment by vEnhance** [comment]
although this is probably the way the code should have been written, i'm a bit hesitant to merge this at a time of year where i expect to be pretty busy. this is a pretty big migration, a lot of code changes (i feel like i'd need to test this a lot before i feel like it's safe to merge) and it would require venueQ to be updated as well

so i'll leave this open for now, but it'll be a while before i get to actually processing it

**Comment by vEnhance** [comment]
okay i've thought about this more and i've decided that fixing this is probably not worth it, due to all the changes that would be needed in many places (in particular, i'm sure we won't catch everything, even with the unit tests). "there's likely some latent bugs here" is a worse outcome to me than a small amount of extra redundancy in the database.

**Comment by Incompleteusern** [comment]
Yeah that makes sense I suppose.

**Pull request closed without merging**